### ISSUE TITLE
Failing extraction on planet.

### ISSUE THREAD

**Issue by jcoupey** [Project-OSRM/osrm-backend#2443]
I'm trying to extract and contract on a HPC cluster, running v5.1.0. Basically I hope to go past the memory requirement limitations to deal with big osm extracts. The drawback is that some parameters are beyond my reach. For example I don't handle the compilation myself and no SWAP is used on the machines.

So far I've been able to work successfully on Europe using 200G of stxxl and 64G RAM. Memory usage topped at 37G for extraction and 19G (only?) for contraction.

I've been trying to extract on a planet file with 96G RAM and up to 400G of stxxl with no luck. The extraction always raises a `std::bad_alloc` after logging `[info] Graph loaded ok and has 621542939 edges`. 
Looking at the code just made me guess that [this comment](https://github.com/Project-OSRM/osrm-backend/blob/5.1/include/util/node_based_graph.hpp#L64-L65) might explain the memory overhead.

I read somewhere on the list that someone managed planet with 88G of RAM, 100G of SWAP and 250G of stxxl so maybe the lack of SWAP is my problem here. On the other hand, my config is quite in line with the description at https://github.com/Project-OSRM/osrm-backend/issues/1845#issuecomment-168059654. And based on the measurements on the wiki (https://github.com/Project-OSRM/osrm-backend/wiki/Data-processing-timings) I should hope to be able to at least extract even if contraction fails afterwards.

 Is there anything I'm missing here, or that I could try?

**Comment by danpat** [comment]
@jcoupey As a reference point, the OSRM demo server runs on an an Amazon `r3.4xlarge` machine.  This has 125GB of RAM, and we have it configured with 250GB of swap, and 250GB for STXXL.

Honestly though, we just threw a bunch of stuff at it, we haven't measured actual usage for a while.  The memory requirements will change depending on which profile you run, the `walk` profile requires significantly more memory/disk than the `car` profile.

**Comment by jcoupey** [comment]
@danpat thanks for the input. I'm using the "less-demanding" default `car` profile. 

What puzzles me the most is that I would expect the overhead to happen during contraction (supposed to be much more memory intensive), not extraction. The fact that extraction required much more memory than contraction on europe might be a clue that something is wrong on my config...

**Comment by TheMarex** [comment]
@jcoupey this changed a little bit with the 4.9.0 release which restructured our pipeline and moved a lot more steps to `osrm-extract`. The recent changes around guidance might have made the memory usage explode there a little bit - we kind of punted on optimizing this part of our stack for efficiency at all.
`osrm-contract` can still have a high memory usage (especially for worst case hierarchies like the pedestrian profile) but for car this is not the bottle neck anymore.

@oxidase has found some problems with broken struct and bit packing that increase memory usage quite a bit - maybe that will yield to a relieve there.

However for the future we already have features in the pipeline that will bump the memory usage. Infamous #77 and eventually #2296 will require more data to stick around. We have adopted a little bit different development style, which is more "ship first, optimize later". I get that this might be frustrating for downstream users that don't really rely on the new features, it is just the was we current set our internal priorities.

To summarize, for now plan to increase the memory limits if you want to process global datasets. I just checked our server logs for 5.1.0 over the last few days. Peak usage for the car profile was 49% of 224GB r3.8xlarge which means 109GB.

**Comment by jcoupey** [comment]
@TheMarex thanks for the update, it does put my observations into perspective. So I guess it all comes down to memory.

I'll see what I can do about it. Alternatively, I might try to break up the extract/contract phase on smaller connected components and serve them separately. But this would require a lot more work on the client side.

**Comment by jcoupey** [comment]
Follow-up: the guys working on the cluster ended up adding a good amount of swap so extraction finally was successful. Currently contracting...

For the record, memory peaked at about 54G of swap on the same machine with 96G RAM.

Thanks again for the input.
